cartographer/mapping/internal/3d/pose_graph_3d.cc

```diff
@@ -509,40 +509,59 @@ void PoseGraph3D::HandleWorkQueue(
 }
 
 void PoseGraph3D::WaitForAllComputations() {
-  bool notification = false;
-  common::MutexLocker locker(&mutex_);
+  int num_trajectory_nodes;
+  {
+    common::MutexLocker locker(&mutex_);
+    num_trajectory_nodes = data_.num_trajectory_nodes;
+  }
+
   const int num_finished_nodes_at_start =
       constraint_builder_.GetNumFinishedNodes();
+
+  // First wait for the work queue to drain so that it's safe to schedule
+  // a WhenDone() callback.
+  {
+    common::MutexLocker locker(&work_queue_mutex_);
+    locker.Await([this]() REQUIRES(work_queue_mutex_) {
+      return work_queue_ == nullptr;
+    });
+  }
+
+  // Now wait for any pending constraint computations to finish.
+  common::Mutex notification_mutex;
```

Comment:
Why not reuse the `mutex_`? Seems this complicates the code quite a bit without a perceivable performance benefit.

Reply:
Done.